### ISSUE TITLE
[Flaky Test] Credentials rotation e2e test stuck in tasks `Deleting old ServiceAccount secrets after rotation of signing key, Removing label from secrets after rotation of ETCD encryption key`

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#6092]
**How to categorize this issue?**

/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:

`E2E Test Suite: [It] Shoot Tests Create Shoot, Rotate CA and Delete Shoot`

**CI link**:

https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6090/pull-gardener-e2e-kind/1534949945625284608
https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6089/pull-gardener-e2e-kind/1534911068743143424
https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6082/pull-gardener-e2e-kind/1534900454847156224

**Reason for failure**:

In all three instances, the test fails in step `Complete credentials rotation` with the shoot stuck in the following state for about 7 min:
```
    {"level":"info","msg":"84%: Shoot State: Processing, Description: Deleting old ServiceAccount secrets after rotation of signing key, Removing label from secrets after rotation of ETCD encryption key","ts":"2022-06-09T17:59:40.274Z"}
```

(With #6088, we know at least in which step it is stuck)
It will be interesting to observe some failing test runs now that we have #6089, which will enable us to check the gardenlet logs for the failed shoot.

**Anything else we need to know**:

The flaky test has been introduced in #6072 (before, only the CAs were rotated).
Note, that this is a different flake compared to #5895 and https://github.com/gardener/gardener/issues/6081.

**Comment by timebertt** [comment]
Another instance reveals the following in the [gardenlet logs](https://storage.googleapis.com/gardener-prow/pr-logs/pull/gardener_gardener/6094/pull-gardener-e2e-kind/1535168125190803456/artifacts/gardener-local-control-plane/containers/gardenlet-8fbd8db58-j9zj2_garden_gardenlet-b5ce0ed1934d4053bf7fece8c7051aba16e444b72b93f052816e930d47f7a57d.log):

```
2022-06-10T08:19:05.224575307Z stderr F {"error":"retry failed with context deadline exceeded, last error: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/apis/apps/v1/namespaces/kube-system/deployments/coredns\": x509: certificate signed by unknown authority","flow":"Shoot cluster reconciliation","level":"error","msg":"Error","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying CoreDNS system component","ts":"2022-06-10T08:19:05.224Z"}
2022-06-10T08:19:05.225652315Z stderr F {"error":"retry failed with context deadline exceeded, last error: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/api/v1/nodes\": x509: certificate signed by unknown authority","flow":"Shoot cluster reconciliation","level":"error","msg":"Error","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying kube-proxy system component","ts":"2022-06-10T08:19:05.225Z"}
```

These are the last logs for shoot `rotate-46lsj`

ref https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6094/pull-gardener-e2e-kind/1535168125190803456#1:build-log.txt%3A1874

**Comment by timebertt** [comment]
The logs of https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6094/pull-gardener-e2e-kind/1535168125190803456 reveal a few interesting aspects.
This might be confusing, but hear me out :)

## Observations

```
... # shoot controller
2022-06-10T08:17:09.549442391Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying Kubernetes API server","ts":"2022-06-10T08:17:09.549Z"}
... # care controller
2022-06-10T08:18:34.376879355Z stderr F {"level":"error","msg":"Could not initialize Shoot client for constraints check: failed to refresh ClientSet's server version: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/version\": x509: certificate signed by unknown authority","shoot":"garden-local/rotate-46lsj","ts":"2022-06-10T08:18:34.376Z"}
2022-06-10T08:18:34.376939385Z stderr F {"level":"error","msg":"Could not initialize Shoot client for garbage collection of shoot garden-local/rotate-46lsj: failed to refresh ClientSet's server version: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/version\": x509: certificate signed by unknown authority","shoot":"garden-local/rotate-46lsj","ts":"2022-06-10T08:18:34.376Z"}
2022-06-10T08:18:34.376946155Z stderr F {"level":"error","msg":"Could not initialize Shoot client for health check: failed to refresh ClientSet's server version: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/version\": x509: certificate signed by unknown authority","shoot":"garden-local/rotate-46lsj","ts":"2022-06-10T08:18:34.376Z"}
... # shoot controller
2022-06-10T08:18:34.655510035Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Waiting until Kubernetes API server rolled out","ts":"2022-06-10T08:18:34.655Z"}
...
2022-06-10T08:18:34.791097325Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying Gardener shoot access resources","ts":"2022-06-10T08:18:34.790Z"}
...
2022-06-10T08:18:34.865232017Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Initializing connection to Shoot","ts":"2022-06-10T08:18:34.865Z"}
... # care controller
2022-06-10T08:19:04.418751151Z stderr F {"level":"info","ts":"2022-06-10T08:19:04.418Z","logger":"clientmap","msg":"Refreshing ClientSet due to changed ClientSetHash","clientmap":"ShootClientMap","key":"garden-local/rotate-46lsj","oldHash":"090930bf18d0672620ab4c4bdcc8dbf3db014105f5c192708122294298aad89a","newHash":"84e4f31c4f1bf90e5339baf6f13943979c76479ad2e881eeaaa5cb0c42ffcd71"}
2022-06-10T08:19:04.41895204Z stderr F {"level":"info","ts":"2022-06-10T08:19:04.418Z","logger":"clientmap","msg":"Invalidating ClientSet","clientmap":"ShootClientMap","key":"garden-local/rotate-46lsj"}
2022-06-10T08:19:04.427559858Z stderr F {"level":"info","ts":"2022-06-10T08:19:04.427Z","logger":"clientmap","msg":"Created new ClientSet","clientmap":"ShootClientMap","key":"garden-local/rotate-46lsj","hash":"84e4f31c4f1bf90e5339baf6f13943979c76479ad2e881eeaaa5cb0c42ffcd71"}
... # shoot controller
2022-06-10T08:19:05.224575307Z stderr F {"error":"retry failed with context deadline exceeded, last error: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/apis/apps/v1/namespaces/kube-system/deployments/coredns\": x509: certificate signed by unknown authority","flow":"Shoot cluster reconciliation","level":"error","msg":"Error","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying CoreDNS system component","ts":"2022-06-10T08:19:05.224Z"}
2022-06-10T08:19:05.225652315Z stderr F {"error":"retry failed with context deadline exceeded, last error: Get \"https://kube-apiserver.shoot--local--rotate-46lsj.svc/api/v1/nodes\": x509: certificate signed by unknown authority","flow":"Shoot cluster reconciliation","level":"error","msg":"Error","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying kube-proxy system component","ts":"2022-06-10T08:19:05.225Z"}
# end of logs for shoot rotate-46lsj
```

During phase `Completing` (started `08:16:28Z`, interrupted by test timeout at `08:26:28Z`), gardenlet got multiple `x509: certificate signed by unknown authority` errors in two situations:

1. In the care controller (health, constraints checks) while the shoot controller was rolling out kube-apiserver (`08:18:34Z`)
2. In the shoot controller while talking to the shoot kube-apiserver in steps `Deploying CoreDNS system component` and following (`08:19:05Z`)

We can also see, that the care controller picked up the new kubeconfig at `08:19:04Z` and never got any certificate errors afterward.

## Race Condition A

These errors should never occur in any phase of the credentials rotation (and we don't see them in successful test runs).
The gardenlet always writes the current CA bundle to the kubeconfig in the `{gardener,gardener-internal}` shoot access secrets in step `Deploying Gardener shoot access resources`.
In step `Initializing connection to Shoot`, the shoot clientmap is supposed to pick up the kubeconfig with the new CA bundle, so that the remaining tasks use it via `botanist.K8sShootClient`.

Here, a first race can occur:
The shoot clientmap reads the kubeconfig secret from a cached client and can thereby still pick up the old CA bundle.
This usually happens, because both steps are executed in quick succession. E.g. in `Completing phase` of the `rotate-46lsj`:
```
2022-06-10T08:18:34.791097325Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying Gardener shoot access resources","ts":"2022-06-10T08:18:34.790Z"}
2022-06-10T08:18:34.865232017Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Initializing connection to Shoot","ts":"2022-06-10T08:18:34.865Z"}
# no logs about refreshing clientset... -> old CA bundle is used during the whole remaining shoot reconciliation
```
Note, that this is not problematic per se. The kubeconfig containing the old CA cert is still "valid", that's the reason why we have the two-phased rotation in the first place...
So if gardenlet picks up the old CA bundle and uses it for the rest of the reconciliation, that's totally fine. In the following reconciliation, the secret cache will be up-to-date and it will use the latest CA bundle. No problem so far.

## Race Condition B

I suspect that there was a second race condition, which could explain the observations.

During phase `Preparing` (started `08:08:28Z`, finished `08:16:28`), gardenlet executed the following two tasks almost simultaneously:

```
2022-06-10T08:12:03.207103786Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Renewing shoot access secrets after creation of new ServiceAccount signing key","ts":"2022-06-10T08:12:03.206Z"}
2022-06-10T08:12:03.214171736Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/rotate-46lsj","task":"Deploying Gardener shoot access resources","ts":"2022-06-10T08:12:03.213Z"}
```

`Renewing shoot access secrets...` patches all shoot access secrets to remove the `token-renew-timestamp` annotation, which will trigger the `token-requestor` to issue new tokens immediately.
`Deploying Gardener shoot access resources` patches gardener's shoot access secrets to update the kubeconfig's CA bundle (decodes the kubeconfig in order to keep the token though).

The problem is, that resource-manager's `token-requestor` uses a patch request without optimistic locking for populating the requested token in the kubeconfig field: https://github.com/gardener/gardener/blob/b1514126925872628f38f0aff2edd66182c36ba7/pkg/resourcemanager/controller/tokenrequestor/reconciler.go#L143-L168

If the `token-requestor` picks up the secret and renews the token while gardenlet is simultaneously writing the new CA bundle, it might happen that the `token-requestor` blindly overwrites the new CA bundle and the old one is kept.

## TL;DR

My current assumption is, that the flake is caused by the following chain of events:

1. During `Preparing`, the race B occurs, and the `gardener` shoot access secret thus contains the old CA bundle. This works fine until the next phase, as the kubeconfig is still "valid".
3. During `Completing`, gardenlet rolls out kube-apiserver and drops the old server certificate. By this, the kubeconfig becomes "invalid" and causes certificate verification errors.
4. Race A occurs and gardenlet's shoot client map doesn't pick up the freshly written new CA bundle. Effectively, gardenlet is still using the old CA cert from before the credentials rotation was started. The shoot client is kept on the botanist for the remaining steps of the reconciliation and hence all shoot API server calls fail.

This also means, that as soon as the shoot reconciliation is retried everything will work as expected. However, due to strict timeouts in our e2e test, it fails before the shoot reconciliation is retried.

**Comment by rfranzke** [comment]
Nice, very interesting findings indeed! :) Thanks for the detailed analysis and the proposed fix 🚀

**Comment by rfranzke** [comment]
There was a new instance of this in https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-e2e-kind/1536668819815141376 (after #6096 was merged).

**Comment by timebertt** [comment]
I checked this test failure and I suspect that it failed because Race Condition B is only partially resolved with https://github.com/gardener/gardener/pull/6096:

In phase `Preparing` these two tasks were again executed in quick succession:
```
2022-06-14T11:43:05.023848348Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/e2e-rotate","task":"Renewing shoot access secrets after creation of new ServiceAccount signing key","ts":"2022-06-14T11:43:05.023Z"}
2022-06-14T11:43:05.037082476Z stderr F {"flow":"Shoot cluster reconciliation","level":"info","msg":"Succeeded","operation":"reconcile","shoot":"garden-local/e2e-rotate","task":"Deploying Gardener shoot access resources","ts":"2022-06-14T11:43:05.036Z"}
```

I suspect, that this time the `token-requestor` was faster than gardenlet and successfully renewed the token as instructed by `Renewing shoot access secrets after creation of new ServiceAccount signing key`.
After this, gardenlet patched the access secret in `Deploying Gardener shoot access resources` in order to write the new CA bundle into the `kubeconfig` field.
However, as https://github.com/gardener/gardener/blob/2b42c1bd6b8302ee40d9a178220f06f9ab5b0cd1/pkg/utils/gardener/shoot.go#L293-L294 doesn't use optimistic locking for the patch, it blindly overwrote the new token.

After this, the next care controller run picked up the changes in the kubeconfig and tried to access the shoot but failed with `401 Unauthorized`:
```
2022-06-14T11:43:31.96550827Z stderr F {"level":"info","ts":"2022-06-14T11:43:31.965Z","logger":"clientmap","msg":"Refreshing ClientSet due to changed ClientSetHash","clientmap":"ShootClientMap","key":"garden-local/e2e-rotate","oldHash":"3d83714239ebf83ba91696c7dc042f4dacaa7d11b1cbec882681ac7b5b11a00c","newHash":"c59284b6b0478dd5dbd985c2de68302e7e946e6b4063741b67f37848f7217fa2"}
2022-06-14T11:43:31.96557007Z stderr F {"level":"info","ts":"2022-06-14T11:43:31.965Z","logger":"clientmap","msg":"Invalidating ClientSet","clientmap":"ShootClientMap","key":"garden-local/e2e-rotate"}
2022-06-14T11:43:31.972918339Z stderr F {"level":"error","msg":"Could not initialize Shoot client for constraints check: error creating new ClientSet for key \"garden-local/e2e-rotate\": error discovering kubernetes version: the server has asked for the client to provide credentials","shoot":"garden-local/e2e-rotate","ts":"2022-06-14T11:43:31.972Z"}
2022-06-14T11:43:31.972935979Z stderr F {"level":"error","msg":"Could not initialize Shoot client for garbage collection of shoot garden-local/e2e-rotate: error creating new ClientSet for key \"garden-local/e2e-rotate\": error discovering kubernetes version: the server has asked for the client to provide credentials","shoot":"garden-local/e2e-rotate","ts":"2022-06-14T11:43:31.972Z"}
2022-06-14T11:43:31.972940769Z stderr F {"level":"error","msg":"Could not initialize Shoot client for health check: error creating new ClientSet for key \"garden-local/e2e-rotate\": error discovering kubernetes version: the server has asked for the client to provide credentials","shoot":"garden-local/e2e-rotate","ts":"2022-06-14T11:43:31.972Z"}
```

Long story short: we should use optimistic locking for patching the shoot access secrets in gardenlet as well.

**Comment by rfranzke** [comment]
This sounds reasonable @timebertt, thanks for checking it out. Will you open a PR?

**Comment by timebertt** [comment]
Sure, I'm on it :)

**Comment by timebertt** [comment]
I have not seen any more instances of this flake since https://github.com/gardener/gardener/pull/6125
/close
